SessionManagerServiceTest. acquireSession_returnsNewSession_whenSessionExpiredAndNotInUse

```
java.lang.AssertionError: Values should be different. Actual: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failEquals(Assert.java:185)
	at org.junit.Assert.assertNotEquals(Assert.java:199)
	at org.junit.Assert.assertNotEquals(Assert.java:211)
	at com.hazelcast.cp.internal.session.AbstractProxySessionManagerTest$11.run(AbstractProxySessionManagerTest.java:236)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1302)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1328)
	at com.hazelcast.cp.internal.session.AbstractProxySessionManagerTest.acquireSession_returnsNewSession_whenSessionExpiredAndNotInUse(AbstractProxySessionManagerTest.java:233)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/547/testReport/junit/com.hazelcast.cp.internal.session/SessionManagerServiceTest/acquireSession_returnsNewSession_whenSessionExpiredAndNotInUse/

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupManager.java
@@ -1001,7 +1001,7 @@ public class MetadataRaftGroupManager implements SnapshotAwareService<MetadataRa
 
         @Override
         public void run() {
-            if (shouldSkipOrReschedule()) {
+            if (shouldRescheduleOrSkip()) {
                 return;
             }
 
@@ -1050,17 +1050,16 @@ public class MetadataRaftGroupManager implements SnapshotAwareService<MetadataRa
             scheduleRaftGroupMembershipManagementTasks();
         }
 
-        private boolean shouldSkipOrReschedule() {
-            if (isDiscoveryCompleted()) {
-                return true;
-            }
+        private boolean shouldRescheduleOrSkip() {
+            // When a node joins to the cluster, first, discoveryCompleted flag is set, then the join flag is set.
+            // Hence, we need to check these flags in the reverse order here.
 
             if (!nodeEngine.getClusterService().isJoined()) {
                 scheduleSelf();
                 return true;
             }
 
-            return false;
+            return isDiscoveryCompleted();
         }
 
         private boolean rescheduleIfCPMemberCountNotSatisfied(Collection<Member> members) {
@@ -1127,6 +1126,7 @@ public class MetadataRaftGroupManager implements SnapshotAwareService<MetadataRa
 
         private void handleDiscoveryFailure() {
             if (terminateOnDiscoveryFailure) {
+                logger.warning("Terminating because of CP discovery failure...");
                 terminateNode();
             } else {
                 logger.warning("Cancelling CP subsystem discovery...");

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/session/AbstractProxySessionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/session/AbstractProxySessionManager.java
@@ -247,12 +247,12 @@ public abstract class AbstractProxySessionManager {
         private final long id;
         private final AtomicInteger acquireCount = new AtomicInteger();
 
+        private final long creationTime;
         private final long ttlMillis;
-        private volatile long accessTime;
 
         SessionState(long id, long ttlMillis) {
             this.id = id;
-            this.accessTime = Clock.currentTimeMillis();
+            this.creationTime = Clock.currentTimeMillis();
             this.ttlMillis = ttlMillis;
         }
 
@@ -265,7 +265,7 @@ public abstract class AbstractProxySessionManager {
         }
 
         private boolean isExpired(long timestamp) {
-            long expirationTime = accessTime + ttlMillis;
+            long expirationTime = creationTime + ttlMillis;
             if (expirationTime < 0) {
                 expirationTime = Long.MAX_VALUE;
             }

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/datastructures/lock/FencedLockAdvancedTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/datastructures/lock/FencedLockAdvancedTest.java
@@ -374,7 +374,10 @@ public class FencedLockAdvancedTest extends HazelcastRaftTestSupport {
             @Override
             public void run() {
                 RaftLockService service = getNodeEngineImpl(lockInstance).getService(RaftLockService.SERVICE_NAME);
-                RaftLock raftLock = service.getRegistryOrNull(lock.getGroupId()).getResourceOrNull(objectName);
+                RaftLockRegistry registry = service.getRegistryOrNull(lock.getGroupId());
+                assertNotNull(registry);
+                RaftLock raftLock = registry.getResourceOrNull(objectName);
+                assertNotNull(raftLock);
                 assertFalse(raftLock.getWaitKeys().isEmpty());
             }
         });

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/session/AbstractProxySessionManagerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/session/AbstractProxySessionManagerTest.java
@@ -233,7 +233,9 @@ public abstract class AbstractProxySessionManagerTest extends HazelcastRaftTestS
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() {
-                assertNotEquals(sessionId, sessionManager.acquireSession(groupId));
+                long newSessionId = sessionManager.acquireSession(groupId);
+                sessionManager.releaseSession(groupId, newSessionId);
+                assertNotEquals(sessionId, newSessionId);
             }
         });
     }

